### ISSUE TITLE
Unable to authenticate with messaging extension

### ISSUE THREAD

**Issue by Blackbaud-ChristiSchneider** [OfficeDev/BotBuilder-MicrosoftTeams#205]
I'm attempting to authenticate with OAuth for my Microsoft Teams messaging extension. It also has a bot, although I don't have that piece really functional yet and am first attempting authentication with the messaging extension. My messaging extension is fully functional if I use my authentication token pasted in from another authenticated website.

I've been working on this for days and have made zero progress. 

The documentation page I found that seems the most promising is [this one](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/messaging-extensions/search-extensions#authentication).

The first thing that I think is unclear is step 2:
> Your service checks whether the user has first authenticated by inspecting the Teams user ID.
What does this mean? I get the user ID and inspect it for what? Some cached authentication that is linked to the key? I assume this would be more clear once I get the other pieces working.

Step 4 is incorrect (or something is missing from the instructions):
> The Microsoft Teams client launches a pop-up window hosting your webpage using the given authentication URL.
When using an `openUrl` action, this opens a new browser tab in the default browser, not a pop-up. 

Step 5 says
> After the user signs in, you should close your window and send an "authentication code" to the Teams client.
This is unclear on how you should get an authentication code, close the window, and how to send the code to Teams.

Is there not a missing step here where something within my service should be storing this authentication code with the user ID mentioned in step 2? This seems like the major missing piece where I cannot figure out what to do.

Step 6
> The Teams client then reissues the query to your service
I have yet to get this to work. I am guessing that this means that it will automatically call the same function that the auth response was returned from originally, although like I said above, I wouldn't expect "inspect the user ID" to do anything different unless there's some code somewhere that's storing the auth code along with the user ID.

Below these summary steps, the documentation goes into more detail.

The [Complete the sign-in flow](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/messaging-extensions/search-extensions#complete-the-sign-in-flow) section states:
> When the sign-in request completes and redirects back to your page, it should perform the following steps:
> 1. Generate a security code. (This can be a random number.) You need to cache this code on your service, along with the credentials obtained through the sign-in flow (such as OAuth 2.0 tokens).
> 2. Call `microsoftTeams.authentication.notifySuccess` and pass the security code.

So I am supposed to generate a security code within my webpage, use the webpage to send this code and my auth token to my service, and call `notifySuccess` with the code? I think the "cache this code and the credentials on your service" should be another step, or I am misunderstanding what is supposed to be happening here.

The next bit says
> At this point, the window closes and control is passed to the Teams client. The client now can reissue the original user query, along with the security code in the state property.
When I call `notifySuccess` from my webpage, the window closes and my app now says "Something went wrong, please try again" in red below the "You'll need to sign in to use this app" line. Nothing happens in my service, including the supposed reissue of the original query.

I have tried a lot of different things to try to have _anything_ happen within my service after the authentication window closes and have been completely unsuccessful thus far. I think this means that I am using `notifySuccess` incorrectly, but it's possible I just haven't guessed the correct listener to try.

My code is based on the [hello world sample](https://github.com/OfficeDev/msteams-samples-hello-world-nodejs), with the basic structure remaining the same (`bot` and `messaging-extension` files). 

My `messaging-extension.js` file, with irrelevant pieces removed/turned into pseudocode:
```
module.exports.setup = function () {
  var builder = require('botbuilder');
  var teamsBuilder = require('botbuilder-teams');
  var bot = require('./bot');
  const https = require('https');

  // This works when I paste an authentication token from somewhere else
  const authToken = bot.connector.settings.authToken;

  bot.connector.onInvoke(function (event, callback) {
    console.log('messaging-extension.onInvoke');
  });

  // Constituent page link
  bot.connector.onAppBasedLinkQuery(function (event, query, callback) {
    console.log('messaging-extension.onAppBasedLinkQuery');
    // TODO check authentication
    let constituentId = 'parsed from link';
    var response = teamsBuilder.ComposeExtensionResponse
      .result('list')
      .attachments([
        getConstituentAttachment(constituentId)
      ])
      .toResponse();

    callback(null, response, 200);
  });

  // Constituent search
  bot.connector.onQuery('constituentSearch', function (event, query, callback) {
    console.log('messaging-extension.constituentSearch');

    // TODO inspect Teams user ID
    // if not authenticated
    let response = teamsBuilder.ComposeExtensionResponse.auth().actions([
        builder.CardAction.openUrl(null,
          'my webpage',
          'Please sign in')
    ]).toResponse();
    // I have also tried the following, which has the same behavior:
    // let response =  {
    //   "composeExtension":{
    //     "type":"auth",
    //     "suggestedActions":{
    //       "actions":[
    //         {
    //           "type": "openUrl",
    //           "value": "my webpage",
    //           "title": "Sign in to this app"
    //         }
    //       ]
    //     }
    //   }
    // };
    callback(null, response, 200);

    // else if authenticated
    // var searchText = 'parsed from query';
    // performConstituentSearch(searchText, callback);
  });

  function getConstituentAttachment(constituentId) {
    // Use the authToken to make an API call to get the constituent information based on the ID
    // Create a card attachment and return it
  }
};
```

`bot.js`:
```
module.exports.setup = function(app) {
  var builder = require('botbuilder');
  var teams = require('botbuilder-teams');
  var config = require('config');

  var connector = new teams.TeamsChatConnector({
    // settings
  });

  let callback = (err, result, statusCode) => {
    console.log('bot.callback');
  };

  connector.onSigninStateVerification((event, query, callback) => {
    console.log('bot.onSigninStateVerification');
  });

  connector.onInvoke((event, callback) => {
    console.log('bot.onInvoke');
  });

  var inMemoryBotStorage = new builder.MemoryBotStorage();

  var bot = new builder.UniversalBot(connector, function(session) {
    console.log('bot.UniversalBot');
  }).set('storage', inMemoryBotStorage);

  bot.on('receive', function (data) {
    console.log('bot.receive');
  });

  app.post('/api/messages', connector.listen());
  app.post('/api/composeExtension', connector.listen());
  module.exports.connector = connector;
};
```

The webpage I redirect to is an angular SPA where I have the user log in and my authentication service call `notifySuccess`:
```
import * as microsoftTeams from '@microsoft/teams-js';

@Injectable()
export class TeamsService {
  constructor(myTokenProvider) {
    microsoftTeams.initialize();
    this.myTokenProvider
      .getDecodedToken({ disableRedirect: true })
      .then((token) => {
        console.log('notifySuccess`);
        // TODO generate an ID and save the token on the server
        let id = 'random id';
        microsoftTeams.authentication.notifySuccess(id);
      }).catch(() => {
        console.log('Not logged in.');
      });
  }
}
```

My response in messaging-extension is based on the [example code](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/messaging-extensions/search-extensions#example-code-in-nodejs), `teamBuilder.ComposeExtensionResponse.auth()`. The code `ComposeExtensionResponse.auth` is found nowhere else on the internet. This example isn't even complete, the code is actually commented out.

In the console logs for my service, I only see `messaging-extension.constituentSearch` one time, before I authenticate. The other console.logs were my attempts at seeing if something else was called when authenticated, but none of those have ever fired.

In my webpage console logs, I see only `notifySuccess`.

Finally, the section [Reissued request example](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/messaging-extensions/search-extensions#reissued-request-example) leads me to believe that _my service_ needs to reissue the request, rather than what is said in step 6, "the Teams client then reissues the query to your service".

I also found the page for [authentication flow for bots](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/authentication/auth-flow-bot), which seems to have a better explanation of what should happen, but I get lost on step 10, the `invoke` message with `name = signin/verifyState` never fires.

I found an old [stackoverflow post](https://stackoverflow.com/questions/41683743/microsoft-teams-desktop-app-does-not-invoke-the-authenticate-successcallback-or) that mentions
> the `microsoftTeams.authentication.authenticate(...)` javascript function as described [here](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/authentication/authentication)
although the link no longer links to a page that describes this function, I believe the link should go somewhere like [this](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/authentication/auth-tab-aad#initiate-authentication-flow). This function is closer to what I would expect to use, which is specifically an authentication popup with a `successCallback` and `failureCallback`. This is client-side code so isn't something you would call from a node.js service, but I expected something similar in the node.js SDK. 

In summary, my problem seems to be that I have no idea how to listen for the authentication response, and there's no clear documentation or examples explaining how to do so.

My setup:
* MacOS Mojave 10.14.6
* Microsoft Teams Version 1.2.00.8866
* Node.js service, using ngrok locally
* botbuilder 3.11.0
* botbuilder-teams 0.2.6

**Comment by Blackbaud-ChristiSchneider** [comment]
More responses that don't work:

```
  let response = teamsBuilder.ComposeExtensionResponse.auth()
    .attachments(new builder.HeroCard()
    .text(`Click below to sign in`)
    .buttons([
      new builder.CardAction()
        .type("signin")
        .value('my webpage')
        .title("Sign in"),
    ])).toResponse();
```

```
  let response = teamsBuilder.ComposeExtensionResponse.auth()
    .attachments([
      new builder.SigninCard()
        .text('Please sign in')
        .button('Sign in', 'my webpage')]
    ).toResponse();
```

```
  let response = teamsBuilder.ComposeExtensionResponse.auth()
    .actions([
      new builder.CardAction()
        .type("signin")
        .value('my webpage')
        .title("Sign in")
    ]).toResponse();
```

**Comment by Blackbaud-ChristiSchneider** [comment]
I also tried running the [auth sample app](https://github.com/OfficeDev/microsoft-teams-sample-auth-node) and have not gotten it to open a sign in window yet. :(

**Comment by Blackbaud-ChristiSchneider** [comment]
I am now trying to get this working in .NET and am running into different issues. I can trigger the sign in flow and receive a token back for my bot through the Bot Framework Emulator, but when I try to use it in Teams, it does not work.

When I test connection in Azure Bot Services, it is successful and returns me a JWT. When I run the bot in Bot Framework Emulator, it runs through the conversation and gives me a token that is a guid + `|livechat`. I ran Fiddler when having this conversation and don't see any requests to `token.botframework.com`

I am using [this sample](https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/18.bot-authentication) and my code is essentially the same. The only thing I changed was to connect to a 3rd party OAuth provider in Azure Bot Services rather than using something like Active Directory.

In Teams, I am able to initiate the conversation and the bot sends back a sign in button, but when I click it, it opens a modal that just says "Bad Request".

In Fiddler I first got a 302 redirect from `https://token.botframework.com/api/oauth/signin?signin=88c4redacted`, then a 400 from `https://token.botframework.com/.auth/web/login/ec36redacted?redirect_uri=https%3a%2f%2ftoken.botframework.com%2fapi%2foauth%2fPostSignInCallback%3fsignin%3d88c4redacted`

I took out the IDs in those urls since I'm not sure what is secret, but can provide them if necessary.

**Comment by clearab** [comment]
Hi - can you please let us know if you've created an app manifest and uploaded your bot to Teams as an app? Authentication will not work if you talk to it directly by GUID.

**Comment by Blackbaud-ChristiSchneider** [comment]
Yes I've got it loaded in Teams and am interacting with it via Teams or Bot Framework Emulator.

**Comment by Wajeed-msft** [comment]
Please try to open your Sign In URL directly in browser. Let us know if you are getting error details in console logs.

**Comment by Blackbaud-ChristiSchneider** [comment]
First error from `https://token.botframework.com/api/oauth/signin?signin=f4bcredacted`
```
[Error] Failed to load resource: the server responded with a status of 400 (Bad Request) (ec36redacted_dd75redacted, line 0)

https://token.botframework.com/.auth/web/login/ec36redacted_dd75redacted?redirect_uri=https%3a%2f%2ftoken.botframework.com%2fapi%2foauth%2fPostSignInCallback%3fsignin%3df4bcredacted
```

then there's a 302 response which redirects to this other url that has no real error
`https://token.botframework.com/.auth/web/login/ec36rredacted_dd75redacted?redirect_uri=https%3a%2f%2ftoken.botframework.com%2fapi%2foauth%2fPostSignInCallback%3fsignin%3df4bcredacted`
```
[Error] Failed to load resource: the server responded with a status of 404 (Not Found) (favicon.ico, line 0)
```

**Comment by Wajeed-msft** [comment]
Could you please share your Bot Id which we could try at our end?

**Comment by Blackbaud-ChristiSchneider** [comment]
`68e416f8-436d-4eed-8706-c57df66723b5`

**Comment by Wajeed-msft** [comment]
I also tried Sign In URL in browser and it's throwing Bad Request. Are you able to login successfully with this bot in Emulator? 
Could you please once try [AuthBot.zip](https://github.com/OfficeDev/microsoft-teams-sample-auth-node/raw/master/manifest/AuthBot.zip) App manifest and see if you are able to login successfully in Microsoft Teams?

**Comment by Blackbaud-ChristiSchneider** [comment]
I am able to login successfully in Bot Framework Emulator, yes, although when I run Fiddler as it's authenticating I don't see any requests to token.botframework.com. I only see `directline` requests on localhost.

I was able to load AuthBot and sign in via google successfully.

**Comment by Blackbaud-ChristiSchneider** [comment]
Well I'm not actually sure I got signed in with AuthBot. I am able to open the modal and log in but it doesn't reply back to the conversation after I do that, and I think it should.

**Comment by Wajeed-msft** [comment]
Can you try opening the login URL sent to Bot Emulator in Browser and see if you get same error?

**Comment by Blackbaud-ChristiSchneider** [comment]
No error on that page, although it's a different url - `https://msteams-authbot.azurewebsites.net/html/auth-start.html?authorizationUrl=https%3A%2F%2Faccounts.google.com%2Fo%2Foauth2%2Fv2%2Fauth%3Fresponse_type%3Dcode%26response_mode%3Dquery%26scope%3Dopenid%2520profile%2520email%26client_id%3D9672redacted.apps.googleusercontent.com%26redirect_uri%3Dhttps%253A%252F%252Fmsteams-authbot.azurewebsites.net%252Fauth%252Fgoogle%252Fcallback%26nonce%3Dab3eredacted%26state%3D%257B%2522securityToken%25redacted&fallbackUrl=https%3A%2F%2Fmsteams-authbot.azurewebsites.net%2Fhtml%2Fauth-start.html%3FauthorizationUrl%3Dhttps%253A%252F%252Faccounts.google.com%252Fo%252Foauth2%252Fv2%252Fauth%253Fresponse_type%253Dcode%2526response_mode%253Dquery%2526scope%253Dopenid%252520profile%252520email%2526client_id%253D9672redacted.apps.googleusercontent.com%2526redirect_uri%253Dhttps%25253A%25252F%25252Fmsteams-authbot.azurewebsites.net%25252Fauth%25252Fgoogle%25252Fcallback%2526nonce%253Dab3eredacted%2526state%253D%25257B%252522securityToken%25redacted`

**Comment by Wajeed-msft** [comment]
Could you please share login URL which you are getting in Bot Emulator for your Bot? Is that URL opening fine in browser?

**Comment by Blackbaud-ChristiSchneider** [comment]
When logging in in Bot Emulator, I don't see any URL in Fiddler (maybe because I logged in previously?)

The pop up asks me if I want to open URL `oauth://SkyApiOAuth&&&68c2redacted|livechat`

The only things I see in Fiddler that look related to the bot are: 
* `http://localhost:56927/v3/directline/conversations/68c2redacted%7Clivechat/activities?watermark=13`
* `https://p39-caldav.icloud.com/mm/sub?token=fea6redacted&key=14redacted` (this doesn't seem related but it shows up every time I hit sign in in the emulator)

**Comment by Blackbaud-ChristiSchneider** [comment]
I've hosted the bot as an Azure App Service if you want to try it locally. `https://snp-skyapi-msteams-svc-temp dot azurewebsites dot net`.

**Comment by Blackbaud-ChristiSchneider** [comment]
I can't connect to the hosted bot in the Bot Emulator for some reason (it returns 500s) but it replies when I use it in Teams.

**Comment by Blackbaud-ChristiSchneider** [comment]
If I'm not mistaken, AuthBot isn't doing what I'm trying to do. It's opening a website (msteams-authbot.azurewebsites.net), which probably uses the `authenticate` and `notifySuccess` javascript functions, which I've also been unable to get working :(

**Comment by Wajeed-msft** [comment]
@Blackbaud-ChristiSchneider - You are correct, didn't realize about the Azure Bot Service part. Could you please try [Leave App manifest](https://github.com/OfficeDev/msteams-sample-line-of-business-apps-csharp/blob/master/Professional%20Services/LeaveBot/Manifest/Leave%20App%20Manifest.zip) which uses Azure Bot Service for authentication. I'll try and get sample code with exact code by tomorrow.

**Comment by Blackbaud-ChristiSchneider** [comment]
Yes, Leave Bot works for me. Thanks.

**Comment by Blackbaud-ChristiSchneider** [comment]
The guid + `|livechat` token I was seeing in Bot Framework Emulator is a fake token that is used when ngrok failed to start, so that was why it looked like it was working in the emulator.

I was able to get the auth to pop up correctly from an `OAuthPrompt` dialog and authenticate after recreating my app in Azure bot services. I believe my original app was created automatically in Teams App Studio, so I'm not sure why it didn't work before.

So, now I'm stuck back in my original problem which is how to get authentication working with a messaging extension. I filed another issue on the teams library - https://github.com/OfficeDev/BotBuilder-MicrosoftTeams-dotnet/issues/41.

**Comment by Blackbaud-ChristiSchneider** [comment]
I found the solution and have gotten OAuth working with messaging extensions. Documenting the solution in the linked issue above.
